src/ray/common/asio/io_context.cc

```diff
@@ -0,0 +1,257 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/asio/io_context.h"
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <utility>
+
+/// C++11 helper for generalized lambda capture.
+template <typename T, typename F>
+class capture_impl {
+  T x;
+  F f;
+
+ public:
+  capture_impl(T &&x, F &&f) : x{std::forward<T>(x)}, f{std::forward<F>(f)} {}
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+
+  template <typename... Ts>
+  auto operator()(Ts &&... args) const -> decltype(f(x, std::forward<Ts>(args)...)) {
+    return f(x, std::forward<Ts>(args)...);
+  }
+};
+
+template <typename T, typename F>
+capture_impl<T, F> capture(T &&x, F &&f) {
+  return capture_impl<T, F>(std::forward<T>(x), std::forward<F>(f));
+}
```

Comment:
This helper allows us to move a given handler into the handler wrapper lambda, avoiding a potentially expensive handler copy, while remaining in C++11 land. With C++14's generalized lambda capture, this helper isn't necessary:
```cpp
auto wrapped_handler = [this, handler = std::move(handler), stats, name, start]() {
  // ...
};
```